pkg/util/rbac/rbac.go

```diff
@@ -68,32 +68,77 @@ func ResourceRoleAssignmentWithName(roleID, spID, resourceType, resourceName, na
+	resourceIdFunction := "resourceId"
+	if subscriptionScope {
+		resourceIdFunction = "subscriptionResourceId"
+	}
 	r := &arm.Resource{
 		Resource: mgmtauthorization.RoleAssignment{
 			Name: to.StringPtr("[" + name + "]"),
 			Type: to.StringPtr("Microsoft.Authorization/roleAssignments"),
 			RoleAssignmentPropertiesWithScope: &mgmtauthorization.RoleAssignmentPropertiesWithScope{
 				Scope:            to.StringPtr("[resourceGroup().id]"),
-				RoleDefinitionID: to.StringPtr("[subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '" + roleID + "')]"),
+				RoleDefinitionID: to.StringPtr("[" + resourceIdFunction + "('Microsoft.Authorization/roleDefinitions', " + roleID + ")]"),
 				PrincipalID:      to.StringPtr("[" + spID + "]"),
 				PrincipalType:    mgmtauthorization.ServicePrincipal,
 			},
 		},
 		APIVersion: azureclient.APIVersion("Microsoft.Authorization"),
+		DependsOn:  dependsOn,
 	}
-
 	if len(condition) > 0 {
 		r.Condition = condition[0]
 	}
+	return r
+}
+
+// ResourceGroupRoleAssignmentWithName returns a Resource granting roleID on the
+// current resource group to spID.  Arguments spID and name must be valid ARM
+// expressions, e.g. "'foo'" or "concat('foo')".  Use this function in ARM
+// templates which have already been deployed, to preserve the name and avoid a
+// RoleAssignmentExists error.
+func ResourceGroupRoleAssignmentWithName(roleID, spID string, name string, condition ...interface{}) *arm.Resource {
+	return resourceGroupRoleAssignmentWithDetails("'"+roleID+"'", spID, name, nil, true, condition...)
+}
+
+// ResourceGroupCustomRoleAssignmentWithName returns a Resource granting roleID on the
+// current resource group to spID.  Arguments spID and name must be valid ARM
+// expressions, e.g. "'foo'" or "concat('foo')".  Use this function in ARM
+// templates which have already been deployed, to preserve the name and avoid a
+// RoleAssignmentExists error. Additionally a dependency is added on the role, use this function
+// when the Role Definition is declared in the same ARM template at non-subscription scope.
+func ResourceGroupCustomRoleAssignment(roleName, spID string, name string, condition ...interface{}) *arm.Resource {
+	dependsOn := []string{
+		"[resourceId('Microsoft.Authorization/roleDefinitions', guid(resourceGroup().id, '" + roleName + "'))]",
+	}
+	return resourceGroupRoleAssignmentWithDetails("guid(resourceGroup().id, '"+roleName+"')", spID, name, dependsOn, false, condition...)
+}
+
+// ResourceRoleDefinition returns a Resource defining a role granting the required
+// actions and dataActions, forbidding notActions and notDataActions. Argument name must be valid ARM expressions.
+func ResourceRoleDefinition(name string, permissions []mgmtauthorization.Permission) *arm.Resource {
+	r := &arm.Resource{
+		Resource: mgmtauthorization.RoleDefinition{
+			Name: to.StringPtr("[" + name + "]"),
```

Comment:
```suggestion
			Name: to.StringPtr("guid(resourceGroup().id, '"+name+"')"),
```
should this be ^ ?  Have I got myself confused?

Reply:
Oh, I think I see.  I think that the user friendly name should be being passed into this function and should be piped to RoleName.  Name should be the generated UUID.